### DISCUSSION TITLE
Async actions in provider setup

### DISCUSSION THREAD

**Discussion by fareeda0** [adonisjs/core#1439] [Help]
Hi all, 

Hope everyone is well 😄 

I see we have transitioned to GitHub discussions now!

Had some questions around performing `async` actions inside of a provider setup.

I am wrapping the `promise-ftp` library inside the `IoC`.

Part of the setup for this lib requires some `async` actions to setup.

Something like:
```ts
const ftpConnection = new PromiseFtp()
await ftpConnection.connect(ftpConfig)
```

I am then using the IoC reference inside of some `ace` commands.

I am able to access the `IoC` from my `ace` commands by adding the following to the command class:

```ts
  public static settings = {
    loadApp: true,
  }
```

However, this is where I run into issues.

It seems `loadApp` only applies to bindings from the `register` method inside the provider. Currently, I am making the bindings inside the `boot` methods, as it is `async` - but then I receive errors running the command as it cannot resolve the name from the `IoC`.

Even with the bindings inside of `boot`, the binding itself has some issues. Using something like:

```ts
  public async boot () {
    // All bindings are ready, feel free to use them
    this.container.singleton('MyOrg/Core/FTP', async () => {
      const ftpConnection = new PromiseFtp()
      const configProvider = this.container.use('Adonis/Core/Config')
      const ftpConfig = configProvider.get('ftp')
      await ftpConnection.connect(ftpConfig)
      return ftpConnection
    })
  }
```

The resulting binding is wrapped inside of a `Promise` and does not allow access to the underlying `ftpConnection` object without awaiting it first. I suspect this is due to adding an `async` before the callback function of `this.container.singleton`

If I move `await ftpConnection.connect(ftpConfig)` to the level above, `this.container.singleton` then it works but doesn't feel like the 'proper' way of doing the binding. 

My questions are:

- What are the recommended ways of performing `async` actions inside of a provider setup?
- How can I access bindings made in the `boot` method inside of my `ace` commands?

**Comment by thetutlage** [comment]
You are using it incorrectly. The factory function `this.container.singleton('MyOrg/Core/FTP', async () => { ` cannot be async. 

Give following code a try and it should work.

```ts
public register () {
  this.container.singleton('MyOrg/Core/FTP', async () => {
    const ftpConnection = new PromiseFtp()
    return ftpConnection
  })
}


public async boot () {
  const FtpConnection = this.container.use('MyOrg/Core/FTP')
  const configProvider = this.container.use('Adonis/Core/Config')
  const ftpConfig = configProvider.get('ftp')
  await FtpConnection.connect(ftpConfig)
}
```

**Reply by fareeda0** [reply]
Thanks @thetutlage 

I had to remove the `async` in 

```ts
public register () {
  this.container.singleton('MyOrg/Core/FTP', async () => {
...
```

From your code above, and it's working now! 🚀
